packages/amazon-sumerian-hosts-core/src/core/awspack/LexUtils.js

```diff
@@ -0,0 +1,112 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+/**
+ * A collection of useful lex related functions.
+ *
+ * @hideconstructor
+ */
+class LexUtils {
+  /**
+   * Downsamples the audio to a target sample rate.
+   *
+   * Inspired by the following blog post from the Lex team:
+   * https://aws.amazon.com/blogs/machine-learning/capturing-voice-input-in-a-browser/
+   * 
+   * @param {Float32Array} buffer - Input audio buffer
+   * @param {float} sourceSampleRate - Sample rate of the input audio buffer
+   * @param {float} targetSampleRate - Sample rate to try to convert to
+   *
+   * @return {Float32Array} Downsampled audio buffer
+   */
+   static downsampleAudio(buffer, sourceSampleRate, targetSampleRate) {
+    if (!buffer || !buffer.length) {
+      return;
+    }
+
+    if (sourceSampleRate === targetSampleRate) {
+      return buffer;
+    }
+
+    if (sourceSampleRate < targetSampleRate) {
+      console.error(`Input Sample rate ${sourceSampleRate} is less than target sample rate ${targetSampleRate}.`);
+      return buffer;
+    }
+
+    const bufferLength = buffer.length;
+    const sampleRateRatio = sourceSampleRate / targetSampleRate;
+    const newLength = Math.round(bufferLength / sampleRateRatio);
+
+    const downsampledBuffer = new Float32Array(newLength);
+
+    let position = 0;
+    let bufferOffset = 0;
+    while (position < newLength) {
+      const nextBufferOffset = Math.round((position + 1) * sampleRateRatio);
+
+      let accumulator = 0;
+      let count = 0;
+      for (let i = bufferOffset; i < nextBufferOffset && i < bufferLength; i++) {
+        accumulator += buffer[i];
+        count++;
+      }
+
+      downsampledBuffer[position] = accumulator / count;
+      position++;
+      bufferOffset = nextBufferOffset;
+    }
+
+    return downsampledBuffer;
+  }
+
+  /**
+   * Converts audio data to WAV.
+   *
+   * Inspired by the following blog post from the Lex team:
+   * https://aws.amazon.com/blogs/machine-learning/capturing-voice-input-in-a-browser/
+   * 
+   * @param {Float32Array} buffer - Input audio buffer
+   * @param {float} targetSampleRate - Sample rate for the output audio
+   *
+   * @return {DataView} Converted audio data
+   */
+  static encodeWAV(buffer, targetSampleRate) {
+    function _writeString(view, offset, string) {
+      for (let i = 0; i < string.length; i++) {
+        view.setUint8(offset + i, string.charCodeAt(i));
+      }
+    }  
+    function _floatTo16BitPCM(view, offset, input) {
+      for (let i = 0; i < input.length; i++, offset += 2) {
+        const s = Math.max(-1, Math.min(1, input[i]));
+        view.setInt16(offset, s < 0 ? s * 0x8000 : s * 0x7fff, true);
+      }
+    }
+
+    if (!buffer) {
+      return;
+    }
+
+    const encodedBuffer = new ArrayBuffer(44 + buffer.length * 2);
+    const view = new DataView(encodedBuffer);
+
+    _writeString(view, 0, 'RIFF');
```

Comment:
I confirmed with Runping that this code comes from the original Sumerian. So while I'm still not satisfied that there's so much unexplained magic, I'm a little less concern since we know this code has been successfully used in production for a while now.